File: day13/src/main.py
from pathlib import Path

def parse_input(filename):
    return [
        (a1, b1, p1, a2, b2, p2)
        for block in Path(filename).read_text().strip().split('\n\n')
        for lines in [block.splitlines()]
        for (a1, a2, b1, b2, p1, p2) in [(
            *map(int, lines[0].split('X+')[1].split(', Y+')),
            *map(int, lines[1].split('X+')[1].split(', Y+')),
            *map(int, lines[2].split('X=')[1].split(', Y='))
        )]
    ]

def solve_system(a1, b1, p1, a2, b2, p2):
    d = a1 * b2 - a2 * b1

    if d == 0 or (p1 * b2 - p2 * b1) % d or (a1 * p2 - a2 * p1) % d:
        return None

    x, y = (p1 * b2 - p2 * b1) // d, (a1 * p2 - a2 * p1) // d

    return (x, y) if x >= 0 and y >= 0 else None

def part1():
    machines, total_cost = parse_input('my_input.txt'), 0

    for i, (a1, b1, p1, a2, b2, p2) in enumerate(machines, 1):
        solution = solve_system(a1, b1, p1, a2, b2, p2)

        if solution:
            total_cost += solution[0] * 3 + solution[1]

        print(f'Progress: {i * 100 // len(machines)}%', end='\r', flush=True)

    print('\r\033[2K', end='\r')

    print(f'Answer: {total_cost}')

def part2():
    machines, total_cost = parse_input('my_input.txt'), 0

    offset = 10_000_000_000_000

    for i, (a1, b1, p1, a2, b2, p2) in enumerate(machines, 1):
        solution = solve_system(a1, b1, p1 + offset, a2, b2, p2 + offset)

        if solution:
            total_cost += solution[0] * 3 + solution[1]

        print(f'Progress: {i * 100 // len(machines)}%', end='\r', flush=True)

    print('\r\033[2K', end='\r')

    print(f'Answer: {total_cost}')

def main():
    part1()
    part2()

if __name__ == '__main__':
    main()

File: day13/src/main.rs
use std::fs;

fn parse_machine(file_name: &str) -> Vec<(i64, i64, i64, i64, i64, i64)> {
    fs::read_to_string(file_name)
        .expect("Unable to read file")
        .trim()
        .split("\n\n")
        .map(|block| {
            let lines: Vec<&str> = block.split('\n').collect();
            let parse_line = |line: &str, x_delim: &str, y_delim: &str| {
                let parts: Vec<i64> = line
                    .split(x_delim).nth(1).unwrap()
                    .split(y_delim).map(|v| v.parse().expect("Invalid number")).collect();
                (parts[0], parts[1])
            };

            let (a1, a2) = parse_line(lines[0], "X+", ", Y+");
            let (b1, b2) = parse_line(lines[1], "X+", ", Y+");
            let (p1, p2) = parse_line(lines[2], "X=", ", Y=");

            (a1, b1, p1, a2, b2, p2)
        })
        .collect()
}

fn solve_system(a1: i64, b1: i64, p1: i64, a2: i64, b2: i64, p2: i64) -> Option<(i64, i64)> {
    let (d, x_num, y_num) = (a1 * b2 - a2 * b1, b2 * p1 - p2 * b1, a1 * p2 - a2 * p1);

    if d == 0 || x_num % d != 0 || y_num % d != 0 {
        None
    } else {
        let (x, y) = (x_num / d, y_num / d);
        (x >= 0 && y >= 0).then_some((x, y))
    }
}

fn part1() {
    let machines = parse_machine("my_input.txt");
    let mut total_cost = 0;

    for (i, &(a1, b1, p1, a2, b2, p2)) in machines.iter().enumerate() {
        if let Some((x, y)) = solve_system(a1, b1, p1, a2, b2, p2) {
            total_cost += x * 3 + y;
        }
        let progress = (i + 1) * 100 / machines.len();
        print!("\rProgress: {}%", progress);
    }

    print!("\r\x1b[2K");
    println!("Answer: {}", total_cost);
}

fn part2() {
    let machines = parse_machine("my_input.txt");
    let mut total_cost = 0;
    let offset = 10_000_000_000_000;

    for (i, &(a1, b1, p1, a2, b2, p2)) in machines.iter().enumerate() {
        if let Some((x, y)) = solve_system(a1, b1, p1 + offset, a2, b2, p2 + offset) {
            total_cost += x * 3 + y;
        }
        let progress = (i + 1) * 100 / machines.len();
        print!("\rProgress: {}%", progress);
    }

    print!("\r\x1b[2K");
    println!("Answer: {}", total_cost);
}

fn main() {
    part1();
    part2();
}
